File: src/php/PhantasyIslandPhpRemoteControl/HttpLayer.php
<?php

namespace PhantasyIslandPhpRemoteControl;

/**
 * 这个文件是 PhantasyIslandPythonRemoteControl 库与仿真平台的核心通信协议部分的 PHP 实现
 */
class HttpLayer
{
    public static function ping()
    {
        return self::send_cmd('ping');
    }

    public static function ping_volatile()
    {
        return self::send_cmd_volatile('ping');
    }

    public static function start()
    {
        return self::send_cmd('start');
    }

    public static function start_volatile()
    {
        return self::send_cmd_volatile('start');
    }

    public static function send_cmd($s)
    {
        $url = 'http://' . Config::$remote_location . '/ECU_HTTP/sendStringCmd?c=' . urlencode($s);
        return self::http_get($url, "send_cmd $s");
    }

    public static function send_cmd_volatile($s)
    {
        $url = 'http://' . Config::$remote_location . '/ECU_HTTP/sendStringCmd?cc=' . urlencode($s);
        return self::http_get($url, "send_cmd_volatile $s");
    }

    public static function get_all_airplane_status()
    {
        $url = 'http://' . Config::$remote_location . '/ECU_HTTP/requestPullAllAirplaneState';
        $result = self::http_get($url, "get_all_airplane_status", 5);
        if ($result === null) {
            throw new \Exception('ConnectionError Cannot Connect to PhantasyIsland, Max retries exceeded.');
        }
        return $result;
    }

    public static function get_airplane_camera_image($port, $camera)
    {
        $url = 'http://' . Config::$remote_location . '/ECU_HTTP/requestPullImage?flyPort=' . $port . '&imageType=' . $camera;
        $j = self::http_get($url, "get_airplane_camera_image", 5);
        if ($j && isset($j['ok']) && $j['ok'] === true) {
            return $j['imgDataString'];
        }
        return null;
    }

    private static function http_get($url, $context_name, $timeout = 10)
    {
        $options = [
            'http' => [
                'method' => "GET",
                'header' => "Accept-language: en\r\n",
                'timeout' => $timeout
            ]
        ];
        $context = stream_context_create($options);
        try {
            $response = @file_get_contents($url, false, $context);
            if ($response === false) {
                error_log("$context_name Error ConnectionError Cannot Connect to PhantasyIsland");
                return ['ok' => false, 'r' => 'ConnectionError Cannot Connect to PhantasyIsland'];
            }
            return json_decode($response, true);
        } catch (\Exception $e) {
            error_log("$context_name Error " . $e->getMessage());
            return ['ok' => false, 'r' => $e->getMessage()];
        }
    }

    public static function process_airplane($j)
    {
        if ($j && isset($j['ok']) && $j['ok'] === true) {
            $airplanes = $j['airplanes'];
            $airplaneStatus = [];
            foreach ($airplanes as $air) {
                $status = [];
                $status['keyName'] = $air['keyName'];
                $status['typeName'] = $air['typeName'];
                $status['updateTimestamp'] = $air['updateTimestamp'];
                $status['status'] = $air['status'];
                
                $camera_front = $air['cameraFront'];
                $status['cameraFront'] = $camera_front['imgDataString'] ?? null;
                
                $camera_down = $air['cameraDown'];
                $status['cameraDown'] = $camera_down['imgDataString'] ?? null;
                
                $airplaneStatus[$status['keyName']] = $status;
            }
            return $airplaneStatus;
        }
        return null;
    }
}

File: src/el/http_layer.erl
-module(http_layer).

-export([ping/0, ping_volatile/0, start/0, start_volatile/0]).
-export([send_cmd/1, send_cmd_volatile/1, get_all_airplane_status/0]).
-export([get_airplane_camera_image/2, process_airplane/1]).

-define(TIMEOUT, 10000).

ping() ->
    send_cmd("ping").

ping_volatile() ->
    send_cmd_volatile("ping").

start() ->
    send_cmd("start").

start_volatile() ->
    send_cmd_volatile("start").

send_cmd(S) ->
    URL = "http://" ++ config:remote_location() ++ "/ECU_HTTP/sendStringCmd?c=" ++ S,
    http_request(URL).

send_cmd_volatile(S) ->
    URL = "http://" ++ config:remote_location() ++ "/ECU_HTTP/sendStringCmd?cc=" ++ S,
    http_request(URL).

get_all_airplane_status() ->
    URL = "http://" ++ config:remote_location() ++ "/ECU_HTTP/requestPullAllAirplaneState",
    case httpc:request(get, {URL, []}, [{timeout, 5000}], []) of
        {ok, {{_, 200, _}, _, Body}} ->
            jsone:decode(list_to_binary(Body));
        {error, Reason} ->
            io:format(standard_error, "ConnectionError Cannot Connect to PhantasyIsland: ~p~n", [Reason]),
            error(connection_error)
    end.

get_airplane_camera_image(Port, Camera) ->
    URL = "http://" ++ config:remote_location() ++ "/ECU_HTTP/requestPullImage?flyPort=" ++ Port ++ "&imageType=" ++ Camera,
    case httpc:request(get, {URL, []}, [{timeout, 5000}], []) of
        {ok, {{_, 200, _}, _, Body}} ->
            J = jsone:decode(list_to_binary(Body)),
            case maps:get(<<"ok">>, J) of
                true -> maps:get(<<"imgDataString">>, J);
                false -> nil
            end;
        {error, _} ->
            nil
    end.

process_airplane(J) ->
    case maps:get(<<"ok">>, J) of
        true ->
            Airplanes = maps:get(<<"airplanes">>, J),
            lists:foldl(fun(Air, Acc) ->
                KeyName = maps:get(<<"keyName">>, Air),
                Status = #{
                    <<"keyName">> => KeyName,
                    <<"typeName">> => maps:get(<<"typeName">>, Air),
                    <<"updateTimestamp">> => maps:get(<<"updateTimestamp">>, Air),
                    <<"status">> => maps:get(<<"status">>, Air),
                    <<"cameraFront">> => get_img_data(maps:get(<<"cameraFront">>, Air)),
                    <<"cameraDown">> => get_img_data(maps:get(<<"cameraDown">>, Air))
                },
                maps:put(KeyName, Status, Acc)
            end, #{}, Airplanes);
        false ->
            nil
    end.

get_img_data(CameraMap) ->
    maps:get(<<"imgDataString">>, CameraMap, nil).

http_request(URL) ->
    case httpc:request(get, {URL, []}, [{timeout, ?TIMEOUT}], []) of
        {ok, {{_, 200, _}, _, Body}} ->
            jsone:decode(list_to_binary(Body));
        {error, timeout} ->
            #{<<"ok">> => false, <<"r">> => <<"Timeout">>};
        {error, _} ->
            #{<<"ok">> => false, <<"r">> => <<"ConnectionError">>}
    end.
